src/Microsoft.AspNetCore.Mvc.Testing/MvcWebApplicationBuilder.cs

```diff
@@ -0,0 +1,137 @@
+        }
+
+        /// <summary>
+        /// Configures <see cref="ApplicationPartManager"/> to include the default set
+        /// of <see cref="ApplicationPart"/> provided by <see cref="DefaultAssemblyPartDiscoveryProvider"/>.
+        /// </summary>
+        /// <returns>An instance of this <see cref="MvcWebApplicationBuilder{TStartup}"/></returns>
+        public MvcWebApplicationBuilder<TStartup> UseApplicationAssemblies()
+        {
+            ApplicationAssemblies.AddRange(DefaultAssemblyPartDiscoveryProvider
+                .DiscoverAssemblyParts(typeof(TStartup).Assembly.GetName().Name)
+                .Select(s => ((AssemblyPart)s).Assembly)
+                .ToList());
+
+            return this;
+        }
+
+        /// <summary>
+        /// Configures the application to use a constant culture for testing purposes.
+        /// </summary>
+        /// <returns>An instance of this <see cref="MvcWebApplicationBuilder{TStartup}"/></returns>
+        public MvcWebApplicationBuilder<TStartup> UseCultureReplacer()
+        {
+            ConfigureAfterStartup(s => s.TryAddEnumerable(ServiceDescriptor.Transient<IStartupFilter, CultureReplacerStartupFilter>()));
+            return this;
+        }
+
+        /// <summary>
+        /// Configures the application content root.
+        /// </summary>
+        /// <param name="solutionName">The glob pattern to use for finding the solution.</param>
+        /// <param name="solutionRelativePath">The relative path to the content root from the solution file.</param>
+        /// <returns>An instance of this <see cref="MvcWebApplicationBuilder{TStartup}"/></returns>
+        public MvcWebApplicationBuilder<TStartup> UseSolutionRelativeContentRoot(
+            string solutionRelativePath,
+            string solutionName = "*.sln")
+        {
+            var applicationBasePath = AppContext.BaseDirectory;
+
+            var directoryInfo = new DirectoryInfo(applicationBasePath);
+            do
+            {
+                var solutionPath = Directory.EnumerateFiles(directoryInfo.FullName, solutionName).FirstOrDefault();
+                if (solutionPath != null)
+                {
+                    ContentRoot = Path.GetFullPath(Path.Combine(directoryInfo.FullName, solutionRelativePath));
+                    return this;
+                }
+
+                directoryInfo = directoryInfo.Parent;
+            }
+            while (directoryInfo.Parent != null);
+
+            throw new Exception($"Solution root could not be located using application root {applicationBasePath}.");
+        }
+
+        public TestServer Build()
+        {
+            using (new CultureReplacer())
+            {
+                var builder = new WebHostBuilder()
+                    .UseContentRoot(ContentRoot)
+                    .ConfigureServices(InitializeServices);
+
+                builder.UseStartup<TestStartup<TStartup>>();
+
+                return new TestServer(builder);
```

Comment:
We could add checks based on whether the user called UseApplicationAssemblies and maybe we can do something to verify the xunit configuration.

There are a few ways of configuring XUnit, not only xunit.runner.json so we can't really tell for sure if shadowCopy (what we care about) is disabled.

Are you ok if we copy xunit.runner.json by default (if the file does not exist on the output) and we provide a property on MSBuild to disable it? We can add a validation in code that checks for xunit.runner.json on the current directory and prints to the console in case its not there or doesn't contain the values that we need. (We can disable the logging to with an environment variable, for the case you go "i know what i'm doing" all the way)